src/binary-reader.c

```diff
@@ -1724,382 +1636,444 @@ static void read_function_body(Context* ctx,
-                            const WasmReadBinaryOptions* options) {
-  LoggingContext logging_context;
-  WASM_ZERO_MEMORY(logging_context);
-  logging_context.reader = reader;
-  logging_context.stream = options->log_stream;
-
-  WasmBinaryReader logging_reader = s_logging_binary_reader;
-  logging_reader.user_data = &logging_context;
+static void read_custom_section(Context* ctx, uint32_t section_size) {
+  WasmStringSlice section_name;
+  in_str(ctx, &section_name, "section name");
+  CALLBACK_CTX(begin_custom_section, section_size, section_name);
 
-  Context context;
-  WASM_ZERO_MEMORY(context);
-  /* all the macros assume a Context* named ctx */
-  Context* ctx = &context;
-  ctx->data = data;
-  ctx->data_size = ctx->section_end = size;
-  ctx->reader = options->log_stream ? &logging_reader : reader;
-  ctx->options = options;
+  WasmBool name_section_ok =
+      ctx->last_section_code >= WASM_BINARY_SECTION_IMPORT;
+  if (ctx->options->read_debug_names && name_section_ok &&
+      strncmp(section_name.start, WASM_BINARY_SECTION_NAME,
+              section_name.length) == 0) {
+    CALLBACK_SECTION(begin_names_section);
+    uint32_t i, num_functions;
+    in_u32_leb128(ctx, &num_functions, "function name count");
+    RAISE_ERROR_UNLESS(num_functions <= num_total_funcs(ctx),
+                       "function name count > function signature count");
+    CALLBACK(on_function_names_count, num_functions);
+    for (i = 0; i < num_functions; ++i) {
+      WasmStringSlice function_name;
+      in_str(ctx, &function_name, "function name");
+      CALLBACK(on_function_name, i, function_name);
 
-  if (setjmp(ctx->error_jmp_buf) == 1) {
-    destroy_context(allocator, ctx);
-    return WASM_ERROR;
+      uint32_t num_locals;
+      in_u32_leb128(ctx, &num_locals, "local name count");
+      CALLBACK(on_local_names_count, i, num_locals);
+      uint32_t j;
+      for (j = 0; j < num_locals; ++j) {
+        WasmStringSlice local_name;
+        in_str(ctx, &local_name, "local name");
+        CALLBACK(on_local_name, i, j, local_name);
+      }
+    }
+    CALLBACK0(end_names_section);
+  } else {
+    /* This is an unknown custom section, skip it. */
+    ctx->offset = ctx->read_end;
   }
+  CALLBACK_CTX0(end_custom_section);
+}
 
-  wasm_reserve_types(allocator, &ctx->param_types,
-                     INITIAL_PARAM_TYPES_CAPACITY);
-  wasm_reserve_uint32s(allocator, &ctx->target_depths,
-                       INITIAL_BR_TABLE_TARGET_CAPACITY);
-
-  uint32_t magic;
-  in_u32(ctx, &magic, "magic");
-  RAISE_ERROR_UNLESS(magic == WASM_BINARY_MAGIC, "bad magic value");
-  uint32_t version;
-  in_u32(ctx, &version, "version");
-  RAISE_ERROR_UNLESS(version == WASM_BINARY_VERSION,
-                     "bad wasm file version: %#x (expected %#x)", version,
-                     WASM_BINARY_VERSION);
-
-  CALLBACK(begin_module, version);
+static void read_type_section(Context* ctx, uint32_t section_size) {
```

Comment:
It gets secretly passed through in the `CALLBACK_SECTION` macro as well, so they all need it.